File: LeetCode/Study Plan/Practice Problems/Medium/Linked List Cycle II/Go/linkedListCycle2.go
package main

import "fmt"

type ListNode struct {
	Val  int
	Next *ListNode
}

func New(val int) *ListNode { return &ListNode{Val: val, Next: nil} }

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func detectCycle(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return nil
	}

	intersectingNode := getIntersectingNode(head)
	if intersectingNode == nil {
		return nil
	}

	p1 := head
	p2 := intersectingNode

	for p1 != p2 {
		p1 = p1.Next
		p2 = p2.Next
	}
	return p1
}

func getIntersectingNode(head *ListNode) *ListNode {
	sp := head
	fp := head

	for fp != nil && fp.Next != nil {
		sp = sp.Next
		fp = fp.Next.Next
		if sp == fp {
			return fp
		}
	}

	return nil
}

func main() {
	head := New(3)
	head.Next = New(2)
	head.Next.Next = New(0)
	head.Next.Next.Next = New(-4)
	head.Next.Next.Next.Next = head.Next
	fmt.Println(detectCycle(head))
}

File: LeetCode/Study Plan/Practice Problems/Medium/Linked List Cycle II/Java/linkedListCycle2.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode detectCycle(ListNode head) {
        if (head == null || head.next == null) return null;

        ListNode intersectingNode = getIntersectingNode(head);
        if (intersectingNode == null) return null;

        ListNode p1 = head;
        ListNode p2 = intersectingNode;
        while (p1 != p2) {
            p1 = p1.next;
            p2 = p2.next;
        } 

        return p1;
    }

    public ListNode getIntersectingNode(ListNode head) {
        ListNode sp = head;
        ListNode fp = head;

        while(fp != null && fp.next != null) {
            sp = sp.next;
            fp = fp.next.next;
            if (sp == fp) return sp;
        }

        return null;
    }
}
